### ISSUE TITLE
Docker fails to create iptables rules?

### ISSUE THREAD

**Issue by aanm** [docker/docker#15948]
I've installed docker with the get.docker.com script on fedora 22 and I have these messages under journalctl:
```
Ago 31 13:53:01 localhost.localdomain docker[1591]: time="2015-08-31T13:53:01.475201320+01:00" level=info msg="Firewalld running: true"
Ago 31 13:53:01 localhost.localdomain firewalld[1232]: 2015-08-31 13:53:01 ERROR: COMMAND_FAILED: '/sbin/iptables -w -t nat -D PREROUTING -m addrtype --dst-type LOCAL -j DOCKER' failed: iptables v1.4.21: Couldn't load target `DOCKER':No s
                                                       
                                                       Try `iptables -h' or 'iptables --help' for more information.
Ago 31 13:53:01 localhost.localdomain firewalld[1232]: 2015-08-31 13:53:01 ERROR: COMMAND_FAILED: '/sbin/iptables -w -t nat -D OUTPUT -m addrtype --dst-type LOCAL ! --dst 127.0.0.0/8 -j DOCKER' failed: iptables v1.4.21: Couldn't load targ
                                                       
                                                       Try `iptables -h' or 'iptables --help' for more information.
Ago 31 13:53:01 localhost.localdomain firewalld[1232]: 2015-08-31 13:53:01 ERROR: COMMAND_FAILED: '/sbin/iptables -w -t nat -D OUTPUT -m addrtype --dst-type LOCAL -j DOCKER' failed: iptables v1.4.21: Couldn't load target `DOCKER':No such 
                                                       
                                                       Try `iptables -h' or 'iptables --help' for more information.
Ago 31 13:53:01 localhost.localdomain firewalld[1232]: 2015-08-31 13:53:01 ERROR: COMMAND_FAILED: '/sbin/iptables -w -t nat -D PREROUTING' failed: iptables: Bad rule (does a matching rule exist in that chain?).
Ago 31 13:53:01 localhost.localdomain firewalld[1232]: 2015-08-31 13:53:01 ERROR: COMMAND_FAILED: '/sbin/iptables -w -t nat -D OUTPUT' failed: iptables: Bad rule (does a matching rule exist in that chain?).
Ago 31 13:53:01 localhost.localdomain firewalld[1232]: 2015-08-31 13:53:01 ERROR: COMMAND_FAILED: '/sbin/iptables -w -t nat -F DOCKER' failed: iptables: No chain/target/match by that name.
Ago 31 13:53:01 localhost.localdomain firewalld[1232]: 2015-08-31 13:53:01 ERROR: COMMAND_FAILED: '/sbin/iptables -w -t nat -X DOCKER' failed: iptables: No chain/target/match by that name.
```

```
docker version
Client:
 Version:      1.8.1
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   d12ea79
 Built:        Thu Aug 13 02:39:27 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.1
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   d12ea79
 Built:        Thu Aug 13 02:39:27 UTC 2015
 OS/Arch:      linux/amd64
```

**Comment by crosbymichael** [comment]
This has been fixed by #16038

thanks!

**Comment by aanm** [comment]
@crosbymichael I'm still having this problem:
```bash
$ docker version 
Client:
 Version:      1.8.2
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   0a8c2e3
 Built:        Thu Sep 10 19:20:17 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.2
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   0a8c2e3
 Built:        Thu Sep 10 19:20:17 UTC 2015
 OS/Arch:      linux/amd64
```

**Comment by calavera** [comment]
@aanm those commits are in master. They will be released in 1.9.

**Comment by aanm** [comment]
@calavera I thought they were released with 1.8.2. Thanks for the clarification then. ;-)

**Comment by danielkza** [comment]
I still see this with Docker 1.9.1 in Fedora 23:

```
$ docker --version
Docker version 1.9.1-fc23, build 110aed2-dirty
```

```
firewalld[22417]: 2015-12-26 10:56:48 ERROR: COMMAND_FAILED: '/sbin/iptables -w -t nat -D OUTPUT -m addrtype --dst-type LOCAL -j DOCKER' failed: iptables: No chain/target/match by that name.
firewalld[22417]: 2015-12-26 10:56:48 ERROR: COMMAND_FAILED: '/sbin/iptables -w -t nat -D PREROUTING' failed: iptables: Bad rule (does a matching rule exist in that chain?).
firewalld[22417]: 2015-12-26 10:56:48 ERROR: COMMAND_FAILED: '/sbin/iptables -w -t nat -D OUTPUT' failed: iptables: Bad rule (does a matching rule exist in that chain?).
firewalld[22417]: 2015-12-26 10:56:48 ERROR: COMMAND_FAILED: '/sbin/iptables -w -t nat -n -L DOCKER' failed: iptables: No chain/target/match by that name.
firewalld[22417]: 2015-12-26 10:56:49 ERROR: COMMAND_FAILED: '/sbin/iptables -w -D FORWARD -i docker0 -o docker0 -j DROP' failed: iptables: Bad rule (does a matching rule exist in that chain?).
firewalld[22417]: 2015-12-26 10:56:49 ERROR: COMMAND_FAILED: '/sbin/iptables -w -t nat -C PREROUTING -m addrtype --dst-type LOCAL -j DOCKER' failed: iptables: No chain/target/match by that name.
firewalld[22417]: 2015-12-26 10:56:49 ERROR: COMMAND_FAILED: '/sbin/iptables -w -t nat -C OUTPUT -m addrtype --dst-type LOCAL -j DOCKER ! --dst 127.0.0.0/8' failed: iptables: No chain/target/match by that name.
```

Strangely running `iptables -L` and `iptables -t nat -L` does show a DOCKER chain.

**Comment by Algram** [comment]
I have exactly the same issue like @danielkza.
```
$ docker --version
Docker version 1.9.1-fc23, build 110aed2-dirty
```

**Comment by thaJeztah** [comment]
@Algram @danielkza I see you're both using Red Hat builds of docker, those contain various modifications, so it could also be related to that. Have you searched the Red Hat issue tracker? https://bugzilla.redhat.com

**Comment by danielkza** [comment]
There are currently only two very small patches in the [F23 tree](http://pkgs.fedoraproject.org/cgit/docker.git/tree/?h=f23), and I have SELinux in permissive mode. Seems quite unlikely to be a distribution issue, specially when it is a know previous bug in vanilla packages.

**Comment by Algram** [comment]
@thaJeztah I have only found these two, which state it has been fixed in 1.9 (https://bugzilla.redhat.com/show_bug.cgi?id=1279002, https://bugzilla.redhat.com/show_bug.cgi?id=1279015).

**Comment by thaJeztah** [comment]
@danielkza there may be more; https://github.com/docker/docker/compare/release/v1.9...projectatomic:fedora-1.9, but I haven't checked 

However, looking at the output you posted earlier, this looks like https://github.com/docker/docker/issues/16137, looks like there's a workaround mentioned in that issue

**Comment by danielkza** [comment]
Do you mean disabling firewalld? It's not an acceptable workaround for me.

**Comment by Algram** [comment]
@danielkza does it work with the firewall disabled?

**Comment by danielkza** [comment]
@Algram Wouldn't disabling firewalld make it impossible to see the issue? The whole problem seems to be related to Docker sending ineffective/wrong commands for firewalld to run.

**Comment by Algram** [comment]
@danielkza e.g. can you access docker containers from inside your network if you turn your firewall off. This is what doesn't work for me.

**Comment by danielkza** [comment]
It only works if I stop firewalld before starting Docker.

**Comment by thaJeztah** [comment]
@danielkza I think there was mention there of the right order in which firewalld had to be started, but I may have read incorrectly (just skimmed over the discussion)

Just tried to reproduce this on a fresh DigitalOcean droplet running Fedora 23, firewalld installed and enabled, installed docker using `curl -sSL https://get.docker.com | sh`, but not able to reproduce :cry: 

Here's the output of that install;

```
Client:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.5.1
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:39:26 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.5.1
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:39:26 UTC 2015
 OS/Arch:      linux/amd64
[root@fedora-1gb-ams3-01 ~]# docker info
Containers: 1
Images: 12
Server Version: 1.9.1
Storage Driver: devicemapper
 Pool Name: docker-252:1-262175-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 107.4 GB
 Backing Filesystem: ext4
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 1.994 GB
 Data Space Total: 107.4 GB
 Data Space Available: 28.9 GB
 Metadata Space Used: 1.946 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.146 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.107 (2015-09-05)
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.2.3-300.fc23.x86_64
Operating System: Fedora 23 (Cloud Edition)
CPUs: 1
Total Memory: 993.5 MiB
Name: fedora-1gb-ams3-01
ID: RKZP:F7O4:YCIN:6GFC:IVYU:PM27:IUC2:B6YW:HQER:IGL6:SIQA:POUJ
```

**Comment by Algram** [comment]
@thaJeztah could you try to start a docker-container? That is when it showed up first for me.

**Comment by thaJeztah** [comment]
@Algram I spun up a number of nginx containers, and they were reachable, no errors. Do you also have the `docker-selinux` package installed (I think that's what it's called in the RedHat repo)?

**Comment by Algram** [comment]
@thaJeztah yes, the docker-selinux-package is installed.

**Comment by danielkza** [comment]
@thaJeztah Docker is already started after firewalld due to the ordering dependency on network.target. I've also tried stopping and starting both manually in different combinations to no avail.

I also have docker-selinux installed, but as I mentioned SELinux is set to permissive system-wide, so it shouldn't be the issue.

Does the DigitalOcean Fedora image actually include firewalld, or did you in install it manually? Did you create any custom rules/zones or anything or the sort?

**Comment by thaJeztah** [comment]
@danielkza yup, it's installed;

```
firewall-cmd --version
0.3.14.2
```

I didn't add any custom rules/zones, it's a clean droplet (in fact, already destroyed the first one, this is a new one). I also just tried the RedHat packages, and no problems either, so, wondering what the difference is here.

FYI this issue https://github.com/docker/docker/issues/15498 seems to indicate that SELinux could still play a role, even in permissive mode.

**Comment by danielkza** [comment]
I'm paying closer attention the error messages now, and I see some things I find strange.

To make sure everything was clean I emptied all the iptables rules, started firewalld with a custom rule to create a DOCKER chain beforehand (to see if it helped), then started docker. These are exactly (and all) the messages I get in the logs:

`` 
firewalld[23101]: ERROR: COMMAND_FAILED: '/sbin/iptables -w -t nat -D PREROUTING -m addrtype --dst-type LOCAL -j DOCKER' failed: iptables: No chain/target/match by that name.
firewalld[23101]: ERROR: COMMAND_FAILED: '/sbin/iptables -w -t nat -D OUTPUT -m addrtype --dst-type LOCAL ! --dst 127.0.0.0/8 -j DOCKER' failed: iptables: No chain/target/match by that name.
firewalld[23101]: ERROR: COMMAND_FAILED: '/sbin/iptables -w -t nat -D OUTPUT -m addrtype --dst-type LOCAL -j DOCKER' failed: iptables: No chain/target/match by that name.
firewalld[23101]: ERROR: COMMAND_FAILED: '/sbin/iptables -w -t nat -D PREROUTING' failed: iptables: Bad rule (does a matching rule exist in that chain?).
firewalld[23101]: ERROR: COMMAND_FAILED: '/sbin/iptables -w -t nat -D OUTPUT' failed: iptables: Bad rule (does a matching rule exist in that chain?).
firewalld[23101]: ERROR: COMMAND_FAILED: '/sbin/iptables -w -t nat -n -L DOCKER' failed: iptables: No chain/target/match by that name.
docker[24160]: time="2015-12-26T21:06:07.562532666-02:00" level=info msg="Default bridge (docker0) is assigned with an IP address 172.17.0.1/16. Daemon option --bip can be used to set a preferred IP address"
firewalld[23101]: ERROR: COMMAND_FAILED: '/sbin/iptables -w -t nat -C POSTROUTING -s 172.17.0.0/16 ! -o docker0 -j MASQUERADE' failed: iptables: No chain/target/match by that name.
firewalld[23101]: ERROR: COMMAND_FAILED: '/sbin/iptables -w -D FORWARD -i docker0 -o docker0 -j DROP' failed: iptables: Bad rule (does a matching rule exist in that chain?).
firewalld[23101]: ERROR: COMMAND_FAILED: '/sbin/iptables -w -t filter -C FORWARD -i docker0 -o docker0 -j ACCEPT' failed: iptables: Bad rule (does a matching rule exist in that chain?).
firewalld[23101]: ERROR: COMMAND_FAILED: '/sbin/iptables -w -t filter -C FORWARD -i docker0 ! -o docker0 -j ACCEPT' failed: iptables: Bad rule (does a matching rule exist in that chain?).
firewalld[23101]: ERROR: COMMAND_FAILED: '/sbin/iptables -w -t filter -C FORWARD -o docker0 -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT' failed: iptables: Bad rule (does a matching rule exist in that chain?).
firewalld[23101]: ERROR: COMMAND_FAILED: '/sbin/iptables -w -t nat -C PREROUTING -m addrtype --dst-type LOCAL -j DOCKER' failed: iptables: No chain/target/match by that name.
firewalld[23101]: ERROR: COMMAND_FAILED: '/sbin/iptables -w -t nat -C OUTPUT -m addrtype --dst-type LOCAL -j DOCKER ! --dst 127.0.0.0/8' failed: iptables: No chain/target/match by that name.
firewalld[23101]: ERROR: COMMAND_FAILED: '/sbin/iptables -w -t filter -C FORWARD -o docker0 -j DOCKER' failed: iptables: No chain/target/match by that name.
```

---

Do these two commands even make sense:

`/sbin/iptables -w -t nat -D PREROUTING`
`/sbin/iptables -w -t nat -D OUTPUT`

They look like trying to delete a rule without any actual rule matching parameters. The DOCKER chain isn't even mentioned! The other commands in the first batch seem to be failing because the rules they specify really don't exist, and therefore can't be deleted.

I don't know enough about Docker and only look superficially at the libnetwork code, but does it try to delete rules straight away without checking if they exist? That will work fine when running the commands directly, but will certainly look like errors in firewalld's eyes.

`/sbin/iptables -w -t nat -n -L DOCKER`

This command works fine when ran manually after starting Docker. Maybe there is a race condition between creating the chain and attempting to use it? I don't know when firewalld returns from the DBus call or if it does any work asynchronously.

**Comment by Algram** [comment]
Just making as much information public as possible. 
This is the log I get:
```

23:11
2015-12-26 23:11:41 ERROR: COMMAND_FAILED: '/sbin/iptables -w -t nat -A DOCKER -p tcp -d 0/0 --dport 25565 -j DNAT --to-destination 172.17.0.3:25565 ! -i docker0' failed: iptables: No chain/target/match by that name. /firewalld
23:01
2015-12-26 23:01:17 ERROR: COMMAND_FAILED: '/sbin/iptables -w -t nat -A PREROUTING -m addrtype --dst-type LOCAL -j DOCKER' failed: iptables v1.4.21: Couldn't load target `DOCKER':No such file or directory Try `iptables -h' or 'iptables --help' for more information. /firewalld
23:01
2015-12-26 23:01:17 ERROR: COMMAND_FAILED: '/sbin/iptables -w -t nat -C PREROUTING -m addrtype --dst-type LOCAL -j DOCKER' failed: iptables v1.4.21: Couldn't load target `DOCKER':No such file or directory Try `iptables -h' or 'iptables --help' for more information. /firewalld
23:01
2015-12-26 23:01:16 ERROR: COMMAND_FAILED: '/sbin/iptables -w -t filter -C FORWARD -o docker0 -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT' failed: iptables: Bad rule (does a matching rule exist in that chain?). /firewalld
23:01
2015-12-26 23:01:16 ERROR: COMMAND_FAILED: '/sbin/iptables -w -t filter -C FORWARD -i docker0 ! -o docker0 -j ACCEPT' failed: iptables: Bad rule (does a matching rule exist in that chain?). /firewalld
23:01
2015-12-26 23:01:16 ERROR: COMMAND_FAILED: '/sbin/iptables -w -t filter -C FORWARD -i docker0 -o docker0 -j ACCEPT' failed: iptables: Bad rule (does a matching rule exist in that chain?). /firewalld
23:01
2015-12-26 23:01:16 ERROR: COMMAND_FAILED: '/sbin/iptables -w -D FORWARD -i docker0 -o docker0 -j DROP' failed: iptables: Bad rule (does a matching rule exist in that chain?). /firewalld
23:01
2015-12-26 23:01:16 ERROR: COMMAND_FAILED: '/sbin/iptables -w -t nat -C POSTROUTING -s 172.17.0.0/16 ! -o docker0 -j MASQUERADE' failed: iptables: No chain/target/match by that name. /firewalld
23:01
Started firewalld - dynamic firewall daemon. systemd
23:01
Starting firewalld - dynamic firewall daemon...
```
It first happened when I tried to install this container: https://hub.docker.com/r/kmb32123/youtube-dl-server/

Note: I am installing the containers over the Fedora Admin-GUI "Cockpit".
Also I have disabled selinux for docker because it doesn't work otherwise since I am on BTRFS.

**Comment by nmcalabroso** [comment]
Still experiencing the issue after the reboot and update on Fedora 23. Same logs as @Algram.

**Comment by nmcalabroso** [comment]
Just want to update you. This StackOverflow [link](http://stackoverflow.com/questions/33600154/docker-not-starting-could-not-delete-the-default-bridge-network-network-bridg/33604859) helped me. Seems like that we just need to reset the network info of docker per update(?).

Also, is it intended that the directory `/var/lib/docker` is accessible only by `root` (and group `root`) instead of the group `docker`?

```bash
# su
# ls -ld /var/lib/docker
# drwx------. 9 root root 4096 Jan  3 04:32 /var/lib/docker/
```

**Comment by danielkza** [comment]
I had already tried erasing `/var/lib/docker/network/files/` previously and it doesn't change anything.

**Comment by Algram** [comment]
Erasing the network files doesn't do the trick for me either.

**Comment by Forestsoft-de** [comment]
The Problem still persist:

`root@vagrant:/vagrant/docker# docker-compose down
Stopping docker_web_1 ... done
Stopping docker_mysql_1 ... done
Stopping docker_redis_1 ... done
Stopping docker_backstop_1 ... done
Removing docker_web_1 ... done
Removing docker_mysql_1 ... done
Removing docker_redis_1 ... done
Removing docker_backstop_1 ... done
Removing network docker_default
root@vagrant:/vagrant/docker# docker-compose up
Creating network "docker_default" with the default driver
ERROR: Failed to program FILTER chain: iptables failed: iptables --wait -I FORWARD -o br-2671a60af486 -j DOCKER: iptables v1.4.21: Couldn't load target `DOCKER':No such file or directory

Try `iptables -h' or 'iptables --help' for more information.
 (exit status 2)
root@vagrant:/vagrant/docker# docker -v
Docker version 1.12.2, build bb80604`

**Comment by thaJeztah** [comment]
@Forestsoft-de what distro are you running on? Do you have firewalld enabled, and was it (re)started after docker was started? If so, then it's likely that firewalld wiped docker's IPTables rules. Restarting the docker daemon should re-create those rules.

**Comment by Forestsoft-de** [comment]
Hey @thaJeztah 

root@vagrant:/vagrant/docker# uname -a
Linux vagrant 3.19.0-25-generic #26~14.04.1-Ubuntu SMP Fri Jul 24 21:16:20 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

Thx for the tip. But why happens this if i run docker-compose down ? :)

**Comment by thaJeztah** [comment]
@Forestsoft-de looks like the error is shown on `docker-compose up`, which creates a network, and exposes/publishes container ports. If the `DOCKER` chain is missing at that point, the error occurs.

If you think it's not related to `firewalld`, please open a new issue

**Comment by Davidrjx** [comment]
i was faced with similar problem, like
```
Creating network "officialha_default" with the default driver
ERROR: Failed to program FILTER chain: COMMAND_FAILED: '/sbin/iptables -w2 -I FORWARD -o br-37873dcbd46b -j DOCKER' failed:
```
docker-compse: 1.8.0/1.20.1
docker-engine: 17.03

it is strange that it was ok before this for docker-compose -f compose.yml up

could anyone have solutions?

**Comment by lordmarkm** [comment]
In case it helps anyone, this exact thing happened to me and apparently our infra team restarted iptables and I had to also restart docker to get it working again.

`sudo service docker restart`

**Comment by audiotecnicality** [comment]
> In case it helps anyone, this exact thing happened to me and apparently our infra team restarted iptables and I had to also restart docker to get it working again.

sudo service docker restart

Restarting the Docker service did the trick for me.

**Comment by Ri0n** [comment]
Is there a solution not involving docker daemon restart?
I just tested 18.09 and hit the problem.

**Comment by henshitou** [comment]
the solution is restart docker daemon as thaJeztah mentioned.
$systemctl restart docker

**Comment by monreal** [comment]
Just ran into this after upgrading to Fedora 32, restarting docker does not seem to fix the problem.

**Comment by tobias-tengler** [comment]
Running into the same issue with SuSEfirewall2 at the moment. Restarting docker correctly creates its iptables rules again.
Once I reboot they are gone again though.
I tried to modify the docker.service to wait for the iptables configuration of SuSEfirewall2:
```sh
[Unit]
After=SuSEfirewall2.service
PartOf=SuSEfirewall2.service
```
but it does not work. Does anyone have an idea?
